File: src/ReetCode/Easy/1to100/Q69/Q69_sqrt.java
/**
* Author    : Lee In Bok 
* Date      : 2023.11.05(Sun)
* Spend Time: 06m 34s
* Runtime   : 2 ms (Beats 91.16%)
* Memory    : 39.5 MB (Beats 64%)
* Algoritm  : Binary Search
 */

public class Q69_sqrt {
    public static void main(String[] args) {
        Solution s = new Solution();

        System.out.println(s.mySqrt(4));
    } 
}

class Solution {
    public int mySqrt(int x) {
        long ans = 0;
        long srt = 0;
        long end = x;

        while(srt <= end){
            long mid = (srt + end) / 2;
            long res = mid * mid;

            if(res == x){
                return (int)mid;
            } else if(res > x){
                end = mid - 1;
            } else {
                srt = mid + 1;
                ans = mid;
            }
        } 

        return (int)ans;
    }
}

File: CodeTest/src/ReetCode/Easy/1to100/Q69/Q69_sqrt.py
"""
# Author    : Lee In Bok 
# Date      : 2023.11.04(Sat)
# Spend Time: 06m 34s
# Runtime   : 42 ms (Beats 70.84%)
# Memory    : 16.3 MB (Beats 32.40%)
# Algoritm  : Binary Search
"""

# class Solution:
#     def mySqrt(self, x: int) -> int:
#         if not x:
#             return 0
        
#         if x < 4:
#             return 1

#         for num in range(x):
#             res = num * num

#             if res == x:
#                 return num
#             elif res > x:
#                 return num - 1
            
class Solution:
    def mySqrt(self, x: int) -> int:        
        ans, srt, end = 0, 0, x

        while srt <= end:
            mid = (srt + end) // 2
            res = mid * mid

            if res == x:
                return mid
            elif res > x:
                end = mid - 1
            else:
                srt = mid + 1
                ans = mid
        return ans
            

s = Solution()
result = s.mySqrt(0)

print(result)